app/index.js

```diff
@@ -267,20 +267,46 @@ server.route({
       let { query } = request;
       let targetYear = null;
 
-      // Check for redis data with this query.
+      // Generate cache key from query params
       const cacheKey = JSON.stringify({ id, query });
-      if (redisEnabled) {
-        const cachedData = await redis.get(cacheKey);
+      if (cache.enabled) {
+        // Load cache entry
+        let cachedData = await redisClient.getObject(cacheKey);
+
+        // If cache entry exists
         if (cachedData) {
-          // Once the data is requested, store for a week
-          await redis.expire(cacheKey, redisCacheTtl);
+          // Check every second if query is running. As database queries can
+          // take a while, this avoids starting the same query in parallel.
+          let duration = 0;
+          let step = 1000;
+          while (cachedData.runningDbQuery) {
+            // If query is running for too long, abort cache load and throw
+            // error.
+            if (duration > cache.loadTimeout) {
+              throw Error(
+                'Database query is taking too long, abort loading cache.'
+              );
+            }
+
+            // Wait
+            await delay(step);
+
+            // Update query duration time
+            duration = duration + step;
 
-          // Inflate stored JSON string
-          const decompressed = pako.inflate(cachedData, { to: 'string' });
+            // Get cached data
+            cachedData = await redisClient.getObject(cacheKey);
+          }
+
+          // Renew cache expiration time
+          await redisClient.expire(cacheKey, cacheTtl);
 
           // Parse string into JSON
-          return JSON.parse(decompressed);
+          return cachedData;
         }
+
+        // No cache data is available, flag runningDbQuery as true
+        await redisClient.setObject(cacheKey, { runningDbQuery: true });
```

Comment:
Capturing notes from conversation with @vgeorge 

The `runningDbQuery` should have a shorter TTL than the resutls, because if the connection is interrupted the variable needs to be cleared so other connections can come through.

Reply:
* I'd generally prefer that environment variables aren't renamed. 
* Please make sure that the error comes through the same as a 503 Gateway timeout error, so that it hits the appropriate code paths in the client. 
* Running Db Query timeout should be on the order of the max database query time, + a factor of safety, so roughly minutes. It should probably be an environment variable, (with a default), so that we can tweak it without a rebuild.